File: Gamblers Ascent/src/Hand.cpp
#include "Hand.h"


Hand::Hand() : isSoft(false), cards()
{

}

void Hand::addCard(Card c)
{
	cards.push_back(c);
}
int Hand::handValue()
{
	int result = 0;
	isSoft = false;
	int numAces = 0;

	for (const Card& card : cards) {
		int cardValue = static_cast<int>(card.rank);

		if (card.rank == Card::rank_t::ace) {
			numAces++;
		}
		else { result += cardValue; }
	}

	// Check for aces and adjust their values if necessary
	while (numAces > 0 && result + 10 <= 21) {
		result += 10;
		numAces--;
	}

	// Update "isSoft" status for the hand
	isSoft = (numAces > 0);

	return result;

}
void Hand::clearHand()
{
	cards.clear();
}

int Hand::cardsInHand()
{
	return cards.size();
}
bool Hand::isPair()
{
	return cards[0].get_value() == cards[1].get_value();
}
bool Hand::isBust()
{
	return (handValue() > 21);
}
bool Hand::isBlackjack()
{
	return (handValue() == 21 || (cards.size() >= 5 && handValue() < 22));
}
// turns isSoft true if the hand is a soft hand. A soft hand is a hand that contains an ace that is counted as 11.
void Hand::updateSoftStatus() {
	isSoft = false; // Reset "isSoft" to false.
	for (const Card& card : cards) {
		if (card.rank == Card::rank_t::ace && handValue() <= 11) {
			isSoft = true; // Set "isSoft" to true if there's an ace with a value of 11.
			break;
		}
	}
}

std::ostream& operator<<(std::ostream& os, Hand& h)
{
	for (int i = 0; i < h.cards.size(); i++)
	{
		os << h.cards[i] << std::endl;
	}
	return os;

}

File: Assets/Scripts/Blackjack/Hand.cs
using System;
using System.Collections;
using System.Collections.Generic;
using UnityEngine;

public class Hand 
{
    public List<Card> Cards { get; } = new List<Card>();
    public bool IsSoft { get; private set; }

    public void AddCard(Card card)
    {
        Cards.Add(card);
    }

    public int HandValue()
    {
        int result = 0;
        IsSoft = false;
        int numAces = 0;

        foreach (var card in Cards)
        {
            int cardValue = card.GetValue();

            if (card.Rank == Rank.Ace)
            {
                numAces++;
            }
            else
            {
                result += cardValue;
            }
        }

        while (numAces > 0)
        {
            if (result + 11 <= 21)
            {
                result += 11;
                numAces--;
            }
            else
            {
                result += 1;
                numAces--;
                IsSoft = true;
            }
        }

        return result;
    }

    public void ClearHand()
    {
        Cards.Clear();
    }

    public bool IsPair()
    {
        return Cards[0].Rank == Cards[1].Rank;
    }

    public bool IsBust()
    {
        return HandValue() > 21;
    }

    public bool IsBlackjack()
    {
        return HandValue() == 21 || (Cards.Count >= 5 && HandValue() < 22);
    }

    public void UpdateSoftStatus()
    {
        IsSoft = false;
        foreach (var card in Cards)
        {
            if (card.Rank == Rank.Ace && HandValue() <= 11)
            {
                IsSoft = true;
                break;
            }
        }
    }

    public override string ToString()
    {
        return string.Join("\n", Cards);
    }

    internal int CardsInHand()
    {
        return Cards.Count;
    }
}

